src/lexitools/commands/correspondences.py

```diff
@@ -31,150 +35,202 @@ def register(parser):
-            delis += [(nA, nB)]
-    G.remove_edges_from(delis)
-
-    table = []
-    for nA, nB, data in sorted(G.edges(data=True), key=lambda x: x[2]['weight'], reverse=True):
-
-        data['language'] = ', '.join(['{0}/{1}'.format(a, b) for a, b in sorted(
-            data['language'])])
-        
-
-        if data['frequency'] >= args.cutoff:
-            table += [[nA, 
-                G.node[nA]['frequency'], 
-                nB, 
-                G.node[nB]['frequency'], 
-                data['frequency'],
-                data['weight'],
-                G.node[nA]['type']]]
-
-    with Table(
-            args, 
-            *['Lang A', 'Freq A', 'Lang B', 'Freq B', 'Occ', 'Weight'], 
-            rows=sorted(table, key=lambda x: (x[-1], x[-2]))):
-        pass
-
-
-    IG = networkx2igraph(G)
-    im = IG.community_infomap()
-    for i, comm in enumerate(im):
-        for node in comm:
-            IG.vs[node]['infomap'] = i+1
-
-    IG.write_gml('{0}-graph.gml'.format(args.dataset))
+        for lA, lB in list(data['frequency']):
+            freq = data['frequency'][(lA, lB)]
+            if freq < args.cutoff:
+                del data['frequency'][(lA, lB)]
+        if len(data['frequency']) == 0:
+            del_edges += [(nA, nB)]
+        else:
+            freq_edge = sum(data['frequency'].values())
+            data['num pairs'] = len(data['frequency'])
+            data['total freq'] = freq_edge
+            data['weight'] = data['num pairs'] / tot_pairs
+    G.remove_edges_from(del_edges)
+    return G
 
+def run(args):
+    ds = get_dataset(args)
+    clts = args.clts.from_config().api
+    if args.model == "bipa":
+        sound_class = clts.transcriptionsystem_dict[args.model]
+    else:
+        sound_class = clts.soundclasses_dict[args.model]
+
+    langgenera_path = "./src/lexitools/commands/lang_genera-v1.0.0.tsv"
+    genera = get_langgenera_mapping(langgenera_path)
+
+    concepticon = Concepticon(Config.from_file().get_clone("concepticon"))
+    concept_dict = concepticon.conceptlists[args.concepts].concepts
+    concepts = {c.concepticon_id for c in concept_dict.values() if
+                c.concepticon_id}
+
+    columns = ('parameter_id', 'concept_name', 'language_id', 'language_name',
+               'value', 'form', 'segments', 'language_glottocode',
+               'concept_concepticon_id', 'language_latitude',
+               'language_longitude')
+    namespace = (('concept_name', 'concept'), ('language_id', 'doculect'),
+                 ('segments', 'tokens'), ('language_glottocode', 'glottolog'),
+                 ('concept_concepticon_id', 'concepticon'),
+                 ('language_latitude',
+                  'latitude'), ('language_longitude', 'longitude'), ('cognacy',
+                                                                     'cognacy'),
+                 ('cogid_cognateset_id', 'cogid'))
+
+    lex = lingpy.LexStat.from_cldf(ds.cldf_dir.joinpath('cldf-metadata.json'),
```

Comment:
I tried with the CLDF asjp dataset instead of beidasinitic... and it filled my entire computer memory just loading the data here... Is there a sort of "lazy loading" or more memory efficient version ?

Surely the rest of the code can be further memory-optimised, for example if I end up with the same problem with networkx, I'll switch to graph-tool (in my experience networkx is good for small to big graphs, but not for very very big ones, for which graph-tool is amazing), but for now I would just be able to load the data.

Reply:
LexStat is not advisable for all of asjp.

Reply:
LexStat has one advantage: for each language pair, it determines the product of each concept slot and also extracts each language itself. However, you can basically also do that "manually" by loading the `forms.csv`. The task here (what is done by lexstat/wordlist in lingpy):

1. assemble forms for each language
2. iterate over each language pair (using itertools.combinations)
3. iterate over all concepts, but if you have n forms in language A for concept X and m forms in language B, go for the product (using itertools.product) between the list of n forms in A and m forms in B

Reply:
As lexstat does all this in a rather save and tested way, I used it for one-dataset-experiments, but generally, the whole code will be faster if we switch to loading the data without lingpy here.

Reply:
Ok, is there another cldf/lexibank approved way I can load data ? I would do it with pandas, but I hear that doing so is not the usual workflow.

Reply:
No pandas, please. Either `pycldf` (@tresoldi, do you have an example?), or you do it as I did it in this one here, with the UnicodeReader: `lexitools/commands/prosody.py`, it will just give you `forms.csv`, but that is reasonably fast, yet you'll have to read in the languages.csv and concepts.csv as well.

Reply:
I would suggest using @LinguList approach in `prosody.py` (below), which is what I have been using: just iterate over rows and collect them if they pass a filter. Passing lambda functions to `pycldf` can take much longer, and if you are used with a `pandas` approach you should find it easier:

```python
    ds = get_dataset(args)
    forms = []
    for row in ds.cldf_reader()['FormTable']:
        if row['Language_ID'] == args.language_id or not args.language_id:
            forms.append(row)
```

Reply:
The alternative is something like:

```
wl = Wordlist.from_cldf(
    metadata_path,
    filter=lambda row: row["language_id"] in args.language_id,
)
```

Reply:
Thanks to both of you, I'll try these out.